internal/metrics/metrics.go

```diff
@@ -0,0 +1,214 @@
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"git.sr.ht/~spc/go-log"
+	"github.com/jakub-dzon/k4e-operator/models"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/tsdb"
+	"io"
+	"os"
+	"path"
+	"sync"
+	"time"
+)
+
+const (
+	millisInHour = 60 * 60 * 1000
+	// DefaultRetentionDuration specifies 7 days data retention
+	DefaultRetentionDuration = int64(7 * 24 * millisInHour)
+	// DefaultMaxBytes specifies no block data size limit
+	DefaultMaxBytes = int64(0)
+)
+
+type DataPoint struct {
+	Time  int64
+	Value float64
+}
+type Series struct {
+	Labels     map[string]string
+	DataPoints []DataPoint
+}
+
+//go:generate mockgen -package=metrics -destination=metrics_api_mock.go . API
+type API interface {
+	io.Closer
+	Deregister() error
+	GetMetricsForTimeRange(tMin time.Time, tMax time.Time) ([]Series, error)
+	AddMetric(value float64, labelsMap map[string]string) error
+}
+
+type TSDB struct {
+	appliedOptions *tsdb.Options
+	db             *tsdb.DB
+	dbLock         sync.RWMutex
+}
+
+func NewTSDB(dataDir string) (*TSDB, error) {
+	metricsDir := path.Join(dataDir, "metrics")
+	if err := os.MkdirAll(metricsDir, 0755); err != nil {
+		return nil, fmt.Errorf("cannot create directory: %w", err)
+	}
+	opts := tsdb.DefaultOptions()
+
+	opts.RetentionDuration = DefaultRetentionDuration
+	opts.MaxBytes = DefaultMaxBytes
+
+	db, err := tsdb.Open(metricsDir, nil, nil, opts, nil)
+	if err != nil {
+		return nil, err
+	}
+	return &TSDB{
+		db:             db,
+		dbLock:         sync.RWMutex{},
+		appliedOptions: opts,
+	}, err
+}
+
+func (t *TSDB) Deregister() error {
+	metricsDir := t.db.Dir()
+	err := t.Close()
+	if err != nil {
+		return err
```

Comment:
We log error in other place in this method. Why not to be consistent and log everywhere?